Hello Nextflow refresh including new modules on containers, config, nf-core, and seqera (#412)
nextflow-io/training: docs/hello_nextflow/06_hello_config.md

    We're going to be playing with that shortly.

!!!note

    Anything you put into the `nextflow.config` can be overridden at runtime by providing the relevant process directives or parameters and values on the command line, or by importing another configuration file, according to the order of precedence described [here](https://www.nextflow.io/docs/latest/config.html).

-   **`demo-params.json`** is a parameter file intended for supplying parameter values to a workflow.
    We will use it in section 5 of this tutorial.

The one thing that's missing is a way to point to the original data without making a copy of it or updating the file paths wherever they're specified.
The simplest solution is to link to the data location.

### 0.2. Create a symbolic link to the data

Run this command from inside the `hello-config` directory:

```bash
ln -s ../data data
```

This creates a symbolic link called `data` pointing to the data directory, which allows us to avoid having to change anything to how the file paths are set up.

```console title="Directory contents"
hello-config
├── data -> ../data
├── demo-params.json

> ├── solutions

Reviewer: This output doesn't match what's in the directory. `solutions` is one level up, in `hello-nextflow`.

Reviewer 2: has been updated

Nextflow supports several different distributed executors, including:

-   HPC (SLURM, PBS, SGE)
-   AWS Batch
-   Google Batch
-   Azure Batch
-   Kubernetes
-   GA4GH TES

The executor is subject to a process directive called `executor`. By default it is set to `local`, so the following configuration is implied:

```groovy title="Built-in configuration"
process {
    executor = 'local'
}
```

Let's look at what it would take to using a Slurm scheduler, assuming we had a connection to a cluster and Slurm was installed appropriately.

!!! warning

    What follows is for demonstration purposes but **will not execute the work** since we don't have access to an external executor.

### 3.1. Set up a Slurm executor

Add the following lines to the `nextflow.config` file:

```groovy title="nextflow.config" linenums="12"
process {
    executor = 'slurm'
}
```

And... that's it! As noted before, this does assume that Slurm itself is already set up for you, but this is really all Nextflow itself needs to know.

Basically we are telling Nextflow to generate a Slurm submission script and submit it using an `sbatch` command.

### 3.2. Launch the workflow to generate the job submission script

> Let's try running this; even though we now it won't execute (since we don't have Slurm set up in the Gitpod environment) we'll be able to see what the submission script looks like.

Reviewer: ```suggestion
Let's try running this; even though we know it won't execute (since we don't have Slurm set up in the Gitpod environment) we'll be able to see what the submission script looks like.
```

    }
    conda_on {
        conda.enabled = true
    }
    local_exec {
        process.executor = 'local'
    }
    slurm_exec {
        process.executor = 'slurm'
    }
}
```

Although it may look like these are going to be mutually exclusive, you can actually combine multiple profiles.
Let's try that now.

### 3.4. Run with a combination of profiles

To use two profiles at the same time, simply give both to the `-profile` parameter, separated by a comma.

```bash
nextflow run main.nf -profile docker_on,local_exec
```

With that, we've returned to the original configuration of using Docker containers with local execution, not that you can tell from the console output:

```console title="Output"
 N E X T F L O W   ~  version 24.02.0-edge

 ┃ Launching `main.nf` [irreverent_bassi] DSL2 - revision: 66cd7c255a

executor >  local (7)
[17/82bbc4] SAMTOOLS_INDEX (2)       [100%] 3 of 3 ✔
[8e/93609c] GATK_HAPLOTYPECALLER (2) [100%] 3 of 3 ✔
[e6/df6740] GATK_JOINTGENOTYPING     [100%] 1 of 1 ✔
```

The point is, we can now use profiles to switch to a different software packaging system (Conda) or a different executor (such as Slurm) with a single command-line option.

> For example, if we were back on our hypothetical HPC from earlier, we would switch to using `-profile conda,slurm` in our Nextflow command line.

Reviewer: ```suggestion
For example, if we were back on our hypothetical HPC from earlier, we would switch to using `-profile conda_on,slurm_exec` in our Nextflow command line.
```

Feel free to test that on your own to satisfy yourself that it works as expected.

> 

Reviewer: ```suggestion

!!! Note
    It may be that you want a profile with settings that are different from nextflow defaults, but you still want to set it automatically without providing it with `-profile`. There's a special profile name for that called `standard`. You can see an example [in the Nextflow official docs](https://www.nextflow.io/docs/latest/config.html#config-profiles).

```

### 4.1. Increase default process resource allocations

By default, Nextflow will use a single CPU and 2GB of memory for each process.
Let's say we decide to double that.

We can modify this behavior by setting the `cpu` and `memory` directives in the `process` block. Add the following to the end of your `nextflow.config` file:

```groovy title="nextflow.config" linenums="20"
process {
    // defaults for all processes
    cpus = 2
    memory = 4.GB
}
```

### 4.2. Run the workflow with the increased defaults

Let's try that out, bearing in mind that we need to keep `-profile my_laptop` in the command going forward.

```bash
nextflow run main.nf -profile my_laptop
```

You may not notice any real difference in how quickly this runs, since this is such a small workload.
But if you have a machine with few CPUs and you allocate a high number per process, you might see process calls getting queued behind each other.
This is because Nextflow will ensure we aren't using more CPUs than are available.

!!! tip

    You can check the number of CPUs allocated to a given process by looking at the `.command.run` log in its work directory.

>     There will be a function called `nxf_launch()` that includes the command `docker run -i --cpu-shares 1024`, where `--cpu-shares` is the number of CPUs given to the process multiplied by 1024.

Reviewer: ```suggestion
    There will be a function called `nxf_launch()` that includes the command `docker run—i—-CPU 1024`, where `--cpu-shares` refers to the CPU time given to this process' tasks. Setting one task's cpu_share to 512 and another to 1024 means that the second task will get double the amount of CPU time as the first.
```

You're probably wondering if you can set resource allocations per individual process, and the answer is of course yes, yes you can!
We'll show you how to do that in a moment.

But first, let's talk about how you can find out how much CPU and memory your processes are likely to need.
The classic approach is to do resource profiling, meaning you run the workflow with some default allocations, record how much each process used, and from there, estimate how to adjust the base allocations.

The truly excellent news on this front is that Nextflow includes built-in tools for doing this, and will happily generate a report for you on request.
Let's try that out.

> ### 4.3. Run the workflow to generate a resource utilization report

Reviewer: Even during the training, we may want to run a command again. Nextflow will throw an error if we're trying to generate a new report with the same name. We can turn on override in the configuration (with `report.overwrite = true`).

Author: Let's see how this plays out in practice during training before deciding if we need to do this. As is, the instructions have folks running it with different filenames each time so maybe not needed.

To have Nextflow generate the report automatically, simply add `-with-report <filename>.html` to your command line.

```bash
nextflow run main.nf -profile my_laptop -with-report report-config-1.html
```

> The report is an html file, which you can download and open in your browser.

Reviewer: ```suggestion
The report is an html file, which you can download and open in your browser. You can also right click it in the file explorer on the left and click on `Show preview` in order to view it on Gitpod.
```

Take a few minutes to look through the report and see if you can identify some opportunities for adjusting resources.
Make sure to click on the tabs that show the utilization results as a percentage of what was allocated.
There is some [documentation](https://www.nextflow.io/docs/latest/reports.html) describing all the available features.

<!-- TODO: insert images -->

One observation is that the `GATK_JOINTGENOTYPING` seems to be very hungry for CPU, which makes sense since it performs a lot of complex calculations.
So we could try boosting that and see if it cuts down on runtime.

However, we seem to have overshot the mark with the memory allocations; all processes are only using a fraction of what we're giving them.
We should dial that back down and save some resources.

### 4.4. Adjust resource allocations for a specific process

> We can specify resource allocations for a given process using the `withName` directive.

Reviewer: ```suggestion
We can specify resource allocations for a given process using the `withName` process selector.
```

    cpus = 2
    memory = 2.GB
    // allocations for a specific process
    withName: 'GATK_JOINTGENOTYPING' {
        cpus = 8
    }
}
```

With that specified, the default settings will apply to all processes **except** the `GATK_JOINTGENOTYPING` process, which is a special snowflake that gets a lot more CPU.
Hopefully that should have an effect.

### 4.5. Run again with the modified configuration

Let's run the workflow again with the modified configuration and with the reporting flag turned on, but notice we're giving the report a different name so we can differentiate them.

```bash
nextflow run main.nf -profile my_laptop -with-report report-config-2.html
```

Once again, you probably won't notice a substantial difference in runtime, because this is such a small workload and the tools spend more time in ancillary tasks than in performing the 'real' work.

However, the second report shows that our resource utilization is more balanced now, and the runtime of the `GATK_JOINTGENOTYPING` process has been cut in half.
We probably didn't need to go all the way to 8 CPUs, but since there's only one call to that process, it's not a huge drain.

<!-- **TODO: screenshots?** -->

> As you can see, this approach is useful when your processes have different resource requirements. It empowers you to can right-size the resource allocations you set up for each process based on actual data, not guesswork.

Reviewer: ```suggestion
As you can see, this approach is useful when your processes have different resource requirements. It empowers you to right-size the resource allocations you set up for each process based on actual data, not guesswork.
```
```suggestion
As you can see, this approach is useful when your processes have different resource requirements. It empowers you to can right-size the resource allocations you set up for each process based on actual data, not guesswork.
```

    The nf-core project has compiled a [collection of configuration files](https://nf-co.re/configs/) shared by various institutions around the world, covering a wide range of HPC and cloud executors.

    Those shared configs are valuable both for people who work there and can therefore just utilize their institution's configuration out of the box, and for people who are looking to develop a configuration for their own infrastructure.

### Takeaway

You know how to allocate process resources, tweak those allocations based on the utilization report, and use a profile to adapt the allocations to the compute environment.

### What's next?

Configuring the parameters destined for the tools and operations wrapped within processes.

---

## 5. Configure workflow parameters

So far we've been exploring options for configuring how Nextflow behaves in terms of executing the work.
That's all well and good, but how do we manage the parameters that are meant for the workflow itself, and the tools it calls within the processes?
That is also something we should be able to do without editing code files every time we want to run on some new data or switch to a different set of reference files.

> As it turns out, there's a lot of overlap between this kind of configuration and the infrastructure configuration, starting with the `nextflow.conf` file, which can also house default values for command line parameters.

Reviewer: ```suggestion
As it turns out, there's a lot of overlap between this kind of configuration and the infrastructure configuration, starting with the `nextflow.config` file, which can also house default values for command line parameters.
```

### 5.1. Move the default parameter declarations to the configuration file

> We originally stored all our default parameter values in the workflow script itself, but we can move them out into the `nextflow.conf` file if we prefer.

Reviewer: ```suggestion
We originally stored all our default parameter values in the workflow script itself, but we can move them out into the `nextflow.config` file if we prefer.
```

    // Accessory files
    reference        = "${projectDir}/data/ref/ref.fasta"
    reference_index  = "${projectDir}/data/ref/ref.fasta.fai"
    reference_dict   = "${projectDir}/data/ref/ref.dict"
    intervals        = "${projectDir}/data/ref/intervals.bed"

    // Base name for final output file
    cohort_name      = "family_trio"
}
```

Feel free to re-run this with the same command as above to verify that it works and still preserves the resumability of the pipeline.

At this point, you may be wondering how to provide actual data and reference files to run this workflow for real, since what we've put in here is just a tiny test set.

There are several options.
As we mentioned earlier (see note at the start of this page), you can override the defaults specified in the `nextflow.config` file by providing directives or parameter values on the command line, or by importing other configuration files.

In this particular case, the best solution is to use a parameter file, which is a JSON file containing key-value pairs for all of the parameters you want to supply values for.

### 5.4. Using a parameter file to override defaults

> We provide a parameter file in the project directory, called `demo-params.json`, which contains key-value pairs for all of the parameters our workflow expects.

Reviewer: ```suggestion
We provide a parameter file in the solutions directory, called `demo-params.json`, which contains key-value pairs for all of the parameters our workflow expects.
```

        docker.enabled = true
    }
    univ_hpc {
        process.executor = 'slurm'
        conda.enabled = true
        process.resourceLimits = [
            memory: 750.GB,
            cpus: 200,
            time: 30.d
        ]
    }
}
```

_After:_

```groovy title="nextflow.config" linenums="3"
profiles {
    docker_on {
        docker.enabled = true
    }
    conda_on {
        conda.enabled = true
    }
    my_laptop {
        process.executor = 'local'
        docker.enabled = true
    }
    univ_hpc {
        process.executor = 'slurm'
        conda.enabled = true
        process.resourceLimits = [
            memory: 750.GB,
            cpus: 200,
            time: 30.d
        ]
    }
    demo {
        // Primary input (file of input files, one per line)
        params.reads_bam        = "data/sample_bams.txt"

        // Accessory files
        params.reference        = "data/ref/ref.fasta"
        params.reference_index  = "data/ref/ref.fasta.fai"
        params.reference_dict   = "data/ref/ref.dict"
        params.intervals        = "data/ref/intervals.bed"

        // Base name for final output file
        params.cohort_name      = "family_trio"
    }
}
```

> As long as we distribute the data bundle with the workflow code, this will enable anyone to quickly try out the workflow without having to supply their own inputs or pointing to the parameter file.

Reviewer: ```suggestion
As long as we distribute the data bundle with the workflow code, this will enable anyone to quickly try out the workflow without having to supply their own inputs or pointing to the parameter file. Besides, we can provide URLs to where files are stored and Nextflow will download them automatically.
```
